Problem: Comments

Almantask opened the pull request:
Issues: too many comments, not enough comments, comments instead of clean code.
Solutions: split code, named constants, add comments.
Example structure for solutions

Please pay the most attention to solution structure and how the code is laid out, not actual insides. 

Resolves issue #2 
Resolves #3

Almantask committed:
Problem: Comments
Issues: too many comments, not enough comments, comments instead of clean code.
Solutions: split code, named constants, add comments.
Example structure for solutions

Almantask committed:
Merge branch 'master' into feature/SolutionStructure

AntipodeYarns commented on General/CodeSmells/Comments/Src/Comments.Problem/Good/MagicValues.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        public double Calculate(float main)
+        {
+            // coeficient * multiplier * taxes - mandatory pay.
```
Following the logic here: 
Coefficient is the first term (1), multiplier is the second (0.99) and taxes is the third (main) and mandatory pay (final term)? Maybe more realistic example values might be better. Also - what is it calculating? Salary? Tax Paid? Not clear.

AntipodeYarns commented on General/CodeSmells/Comments/Src/Comments.Problem/Good/NotEnoughComments.cs:
```diff
@@ -0,0 +1,71 @@
+            {
+                right.Add(unsorted[i]);
+            }
+
```
Wouldn't this be better in a single while loop? 
int ctr = 0
while { left.Count + right.Count <= unsorted.Count }
{
    left.Add(i);
    if (unsorted.Count % 2 == 0 && i <= unsorted.Count / 2)
    right.Add(unsorted.Count(i)
}

AntipodeYarns commented on General/CodeSmells/Comments/Src/Comments.Problem/Good/WarehouseDivider.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace Comments.Problem.Good
+{
```
Where is the Bad for this problem?

AntipodeYarns reviewed (commented):
Added some question inline with individual files. Please review

Almantask committed:
Magic values and comments problem solution

Almantask committed:
Comments problem solved
Added good and bad examples for too many comments (booze distributor)
Added tests for every problem
Added more remarks in comments for every problem
Changed solution layout to be:  problem/sub-problem/example

AntipodeYarns reviewed (dismissed):
Looks fine to me.

joshmond commented:
Good job, all looks fine to me

ellmaster commented:
General structure Comments.Problem/X/ProblemA/ExampleA.cs looks fine where X = Good/Bad but I am concerned when it grows.
Is Comments.Problem/ProblemA/X/ExampleA.cs a better structure?
Is Comments.Problem/ProblemA/ExampleA/X.cs a alternative to take into consideration?
My personal favorite would be Comments.Problem/ProblemA/ExampleA.cs (partial(?) with private(?) nested classes Good and Bad).

First thoughts:
Case 1:
Splitting up into namespaces good and bad will make those namespaces quite big. Additionally, will there be a case where you will not need both of those large namespaces?
Case 2:
ProblemA makes the namespace itself very small and X divides it unnecessarily further.
Case 3:
Makes it hard to use more than two examples at once.
Case 4:
Is not the best option in terms of using nested classes but of all the options it seems to be most usable for what it's planned to do.

Almantask commented:
@ellmaster 

> General structure Comments.Problem/X/ProblemA/ExampleA.cs looks fine where X = Good/Bad but I am concerned when it grows.

What concerns do you have? 
Actually I have considered growth, that's why it is what it is. The structure is not exactly what you wrote. 
It's:
ProblemName.Problem/X/ProblemVariation/...
...- can be many files to illustrate a problem variation. 

So instead of: 
```
Good/VariationX/...
Bad/VariationX...
```
alternative would be:
```
VariantionX/Good/...
VariantionY/Bad/...
```
Which makes sense I guess... I thought at first that I would be duplicating Good and Bad namespaces across, but now I duplicate VariationX across anyways.  Though I duplicate variations 2 times and Good/Bad folders will be duplicated 3 times. Logical grouping makes sense.

@ellmaster will refactor :)

Almantask commented:
@ellmaster 

> My personal favorite would be Comments.Problem/ProblemA/ExampleA.cs (partial(?) with private(?) nested classes Good and Bad).

It needs to be testable, easy to read and understand, conventional. Partial classes are mostly used for autogenerated stuff. Also, examples can be multiple files, so it would be convoluted. Nested classes for good and bad would be too much because I don't think anyone would want to see bad when they look at good (I want to limit the scope I have to look at once). Lastly, people would like to have a side by side example. Possible with partial classes, but again it's not great scalability with more complicated examples.

Almantask commented:
Case 1:

> Splitting up into namespaces good and bad will make those namespaces quite big. Additionally, will there be a case where you will not need both of those large namespaces?

It won't go deeper than this
```
Good/Variation1/File11.cs
Good/Variation2/File12.cs
```
```
Good/Variation2/File21.cs
Good/Variation2/File22
```
So consider VariationX as a small component with all the needed scripts at the root level. Should never be more than 10.

There will always be good and bad solutions, to illustrate a problem(misconception, wrong implementation, maintainability problem, performance problems or similar)  and solution.

Good and bad namespaces are needed because names for good and bad will usually match, so namespaces will be the separating factor.

For simple cases we could have done something like:
```cs
Variation1/Variation1Good.cs
Variation1/Variation1Bad.cs
```
but if there is a need to scale it, we will need to add folders not to mix good with bad, we will need to change names... And if that's the case we might as well **keep it consistent**.

Almantask committed:
Change problem structure
from
Good/VariationX/...
Bad/VariationX/...
to
VariationX/Good/...
VariationX/Bad/...

Almantask committed:
Naming made consistent

TheCodemonger commented on General/CodeSmells/Comments/Src/Comments.Problem/TooManyComments/Good/Order.cs:
```diff
@@ -0,0 +1,21 @@
+﻿namespace Comments.Problem.TooManyComments.Good
+{
+    // We work with orders and people, so we might as well make concepts for each.
```
XML Summary?

Almantask replied:
Might as well, yes :)

Almantask replied:
Might as well, yes :)

Almantask replied:
Added XML docs.

TheCodemonger commented on General/CodeSmells/Comments/Src/Comments.Problem/TooManyComments/Good/OrderStatus.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace Comments.Problem.TooManyComments.Good
+{
+    // For a defined set of values prefer to use a set of constants or better- enum.
```
XML Summary?

Almantask replied:
Might as well, yes :)

Almantask replied:
Added XML docs.

TheCodemonger commented on General/CodeSmells/Comments/Src/Comments.Problem/TooManyComments/Good/Person.cs:
```diff
@@ -0,0 +1,21 @@
+        public DateTime Birthday { get; }
+        // If objects can manage their own state without outside intervention- they should.
+        // Person has enough info to calculate their own age.
+        public int Age => (int)(DateTime.Now - Birthday).TotalDays / 365;
```
public int Age => (int)Math.Floor((DateTime.Now - Birthday).TotalDays / 365.25);// Because leap years.

Almantask replied:
Holy crap 🗡 Will change

Almantask replied:
Changed to include leap years.

TheCodemonger reviewed (commented):
Can't comment on the tests, not sufficiently experienced working with them. Seems fine otherwise.

Almantask committed:
Fixed based on remarks
Fixed age calculation with  leap years in consideration
Added xml docs for Order and OrderStatus

Almantask committed:
Gave more clear names to problem variations

Almantask committed:
Namespace naming adjusted

TheCodemonger reviewed (approved):
Looks good. 
A couple files were flagged as changed, but didn't appear to be changed.

ellmaster commented:
@Almantask 
Yeah you're right. I did it kind of on the fly and wasn't aware of the true nature of the code structure

I guess the structure is fine now, at least I have no further concerns about what I see. Therefor I approve! :)
